File: vbr/backend/vehiclebreakdown/src/main/java/com/vehiclebreakdown/service/BookingService.java
package com.vehiclebreakdown.service;

import com.vehiclebreakdown.dto.BookingRequestDto;
import com.vehiclebreakdown.model.Booking;

import java.util.List;
import java.util.Optional;

public interface BookingService {
    Booking createBooking(BookingRequestDto bookingDto);
    Optional<Booking> getBookingById(Long id);
    List<Booking> getBookingsByUserId(Long userId);
    List<Booking> getBookingsByMechanicId(Long mechanicId);
    List<Booking> getPendingBookingsForMechanic(Long mechanicId);
    Booking acceptBooking(Long bookingId, Long mechanicId);
    Booking rejectBooking(Long bookingId, String reason);

    Booking updateBookingStatus(Long bookingId, Booking.BookingStatus status);
    List<Booking> getAllBookings();
    void cancelBooking(Long bookingId);
    List<Booking> getBookingsByStatus(Booking.BookingStatus status);
}

File: vbr/frontend/src/app/services/booking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Booking, BookingRequest, BookingStatus } from '../models/booking.model';

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  createBooking(bookingData: BookingRequest): Observable<any> {
    return this.http.post(`${this.apiUrl}/bookings`, bookingData);
  }

  getBookingById(id: number): Observable<Booking> {
    return this.http.get<Booking>(`${this.apiUrl}/bookings/${id}`);
  }

  getUserBookings(userId: number): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/bookings/user/${userId}`);
  }

  getMechanicBookings(mechanicId: number): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/bookings/mechanic/${mechanicId}`);
  }

  acceptBooking(bookingId: number, mechanicId: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/bookings/${bookingId}/accept`, { mechanicId });
  }

  // rejectBooking(bookingId: number): Observable<any> {
  //   return this.http.post(`${this.apiUrl}/bookings/${bookingId}/reject`, {});
  // }
  // Accepts the reason as a parameter and sends it in the request body
rejectBooking(bookingId: number, reason: string): Observable<any> {
  return this.http.post(`${this.apiUrl}/bookings/${bookingId}/reject`, { reason });
}




  updateBookingStatus(bookingId: number, status: BookingStatus): Observable<any> {
    return this.http.put(`${this.apiUrl}/bookings/${bookingId}/status`, { status });
  }

  cancelBooking(bookingId: number): Observable<any> {
    return this.updateBookingStatus(bookingId, BookingStatus.CANCELLED);
  }
}
